### PR TITLE
test(governance): vitest bootstrap + harden execution-code calldata patterns

### LINKED ISSUE #400: fix(governance): harden execution-code calldata decoding against malformed proposals
Part of the 2026-07 repository audit — M1 — Critical correctness & security.

The proposal execution-code viewer is the screen voters use to verify what a governance proposal will actually do on-chain. Anyone can create a proposal, so its inputs are adversarial by definition. The pattern files that translate decoded calldata into human-readable summaries contain **26 non-null assertions** (`arg!.value`) on decoded arguments. Today their only safety net is a hand-maintained `requiredArgsCount` per pattern plus a catch-all that renders literal `"Error: Failed to process transfer call"` strings into the voter-facing summary **labelled `confidence: "high"`** — instead of falling back to the generic/raw-calldata rendering the dispatcher already has. Any drift between a pattern's destructuring and its `requiredArgsCount` reintroduces a swallowed `TypeError` and a misleading summary on the one screen that must never mislead.

## Current behavior (evidence)

**Note on severity (verified 2026-07-03):** the original audit brief said malformed args "throw a TypeError during render". Re-verification shows `createPattern` catches the throw (see below), so there is no hard white-screen today from these exact lines. The real defects are: (a) misleading `Error: ...` strings rendered as high-confidence summaries, (b) a fragile invariant (`requiredArgsCount` must match each destructuring) enforced by neither the compiler nor tests, and (c) no error boundary if anything in the viewer does throw during render. Implement the hardening as specified; do not hunt for a crash repro.

**26 non-null assertions across the three files** (count verified with `grep -o '!\.' <file> | wc -l`):

- `apps/governance.mento.org/app/components/proposal/execution-code/patterns/token-patterns.ts` — **15** (lines 13, 16, 29, 32, 45, 48, 63, 77, 78, 79 ×2, 88, 89, 102, 114)
- `apps/governance.mento.org/app/components/proposal/execution-code/patterns/proxy-patterns.ts` — **7** (lines 12, 13, 23, 24, 34, 35, 52)
- `apps/governance.mento.org/app/components/proposal/execution-code/patterns/oracle-patterns.ts` — **4** (lines 14, 17, 30, 33)

Representative excerpt, `token-patterns.ts:10-24`:

```ts
  "transfer(address,uint256)": createPattern(
    (contract, args) => {
      const [recipient, amount] = args;
      const recipientName = getAddressNameFromCache(String(recipient!.value));
      const token = getContractInfo(contract.address);
      const formattedAmount = formatTokenAmount(
        String(amount!.value),
        token?.decimals || DEFAULT_TOKEN_DECIMALS,
      );

      return `Send ${formattedAmount} ${token?.symbol || "tokens"} to ${recipientName}`;
    },
    2,
    "transfer",
  ),
```

The assertions are needed only because `noUncheckedIndexedAccess: true` is on (`packages/typescript-config/base.json:13`), which correctly types each destructured element as `DecodedArg | undefined` — the `!` suppresses exactly the check we want.

The wrapper converts failures into voter-visible error strings, `patterns/base-pattern.ts:34-58`:

```ts
      if (requiredArgsCount !== undefined && functionName) {
        if (args.length < requiredArgsCount) {
          return `Error: ${functionName} requires at least ${requiredArgsCount} arguments, got ${args.length}`;
        }
        // ... per-arg null check returning `Error: Missing required argument ...`
      }

      return fn(contract, args, value);
    } catch (error) {
      console.error(
        `Pattern execution error for ${functionName || "unknown function"}:`,
        error,
      );
      return `Error: Failed to process ${functionName || "function"} call`;
```

The dispatcher renders whatever the pattern returns as a high-confidence summary, `transaction-translator.ts:80-92`:

```ts
    const pattern = patternManager.getPattern(decoded.functionSignature);
    if (pattern) {
      const contractInfo = { address: transaction.address };
      const description = pattern(
        contractInfo,
        decoded.args || [],
        transaction.value,
      );
      return {
        description,
        confidence: "high",
      };
    }
```

The generic fallback path already exists directly below (`transaction-translator.ts:94-113`): `Call ${decoded.functionName} on ${...}` with `confidence: "medium"`. `SimpleView.tsx:28-37` renders `summary.description` verbatim, so the `Error: ...` strings reach voters.

Pattern return type is currently `string`, `patterns/types.ts:13-15`:

```ts
export interface PatternFunction {
  (contract: ContractInfo, args: DecodedArg[], value: string | number): string;
}
```

The app already has an error boundary component in use — `app/components/providers.tsx:9` imports `ErrorBoundary` from `@sentry/nextjs` and wraps the tree at lines 38-45. `ExecutionCode.tsx` (a `"use client"` component, line 1) does not use one; it already imports `RawView` (line 21), which renders `JSON.stringify(transactions)` and is a ready-made raw-calldata fallback.

The governance app has **no vitest setup**: `apps/governance.mento.org/package.json` has no `test` script and no vitest devDependency. `apps/app.mento.org/vitest.config.ts` is the config pattern to copy; `@repo/vitest-config` exposes `./shared` (`packages/vitest-config/vitest.shared.js`), and the root `turbo.json:35-38` already defines the `test` task.

## Required change

1. **Change the pattern contract to allow a typed "unrecognized" result.** In `patterns/types.ts`, change `PatternFunction` to return `string | null`. In `patterns/base-pattern.ts`, change `StandardPatternFunction` and the `fn` parameter of `createPattern` to return `string | null`, and replace all three `return \`Error: ...\`` sites with `return null` (keep the `console.error` in the catch). Null means "shape did not match — fall back".
2. **Make the dispatcher fall through on null.** In `transaction-translator.ts` (`translateDecodedTransaction`, lines 80-92): when the pattern returns `null`, do not return — fall through to the existing generic ABI-based description below (confidence `"medium"`). Sketch — adapt to local style:

   ```ts
   if (pattern) {
     const description = pattern(contractInfo, decoded.args || [], transaction.value);
     if (description !== null) {
       return { description, confidence: "high" };
     }
     // fall through to generic description
   }
   ```

3. **Remove all 26 `!` assertions** in `token-patterns.ts`, `proxy-patterns.ts`, `oracle-patterns.ts`. Replace with explicit guards immediately after destructuring; `noUncheckedIndexedAccess` will then enforce the guard at compile time. Sketch — adapt to local style:

   ```ts
   const [recipient, amount] = args;
   if (recipient === undefined || amount === undefined) return null;
   const recipientName = getAddressNameFromCache(String(recipient.value));
   ```

   Do this per pattern function (8 in token, 4 in proxy, 2 in oracle). Guard every argument the function actually reads (e.g. `lock(...)` reads `account`, `amount`, `slopePeriod`, `cliff`).
4. **Bootstrap vitest for the governance app** (the governance-tests bootstrap issue has not landed; these are plain pure-function tests). Create `apps/governance.mento.org/vitest.config.ts` copying the `apps/app.mento.org/vitest.config.ts` pattern but minimal: merge `@repo/vitest-config/shared`, alias `"@" → ./app`, and override `test.include` to `["app/**/*.test.ts", "app/**/*.test.tsx"]`. No jsdom, no jsx settings — node environment (the shared default) suffices. Add to `apps/governance.mento.org/package.json`: script `"test": "vitest run"`, devDependencies `"@repo/vitest-config": "workspace:*"` and `"vitest": "catalog:"` (both already exist in the workspace — this is not a new dependency). Run `pnpm install` from the repo root afterwards.
5. **Add unit tests** colocated in the patterns directory: `token-patterns.test.ts`, `proxy-patterns.test.ts`, `oracle-patterns.test.ts`. For each pattern function: malformed-args cases (empty `args` array; too few args; an arg whose `value` is `null` cast via `unknown`) asserting the wrapped function returns `null`; plus one happy-path test per file pinning a real description (e.g. `transfer` with `value: "1000000000000000000"` produces a `Send ... to ...` string). Mock the address resolver at the top of each test file (same module specifier the source files use):

   ```ts
   vi.mock("../../services/address-resolver-service", () => ({
     getAddressNameFromCache: (address: string) => address,
     getContractInfo: () => undefined,
     addressResolverService: {
       resolveFromCacheWithContext: (address: string) => ({ name: address }),
     },
   }));
   ```

   Follow the existing repo test style (behaviour assertions, `vi.mock` — see `packages/web3/src/features/borrow/hooks/use-open-trove.test.ts`).
6. **Defence-in-depth error boundary.** In `ExecutionCode.tsx`, wrap the `<Tabs>` block (lines 58-86) in the `ErrorBoundary` from `@sentry/nextjs` (already a dependency, already used in `providers.tsx`) with a `fallback` that renders a short "Could not decode this proposal's execution code — raw calldata shown below." notice plus `<RawView transactions={transactions} />`.

## Out of scope

- The **27 identical assertions** in the sibling files `governance-patterns.ts` (5), `reserve-patterns.ts` (19), `protocol-patterns.ts` (3). Steps 1-2 already give those files the null-fallback safety net at runtime; removing their `!`s is the same mechanical treatment and should be a follow-up, not smuggled into this diff.
- Fixing the stale CLAUDE.md guidance (React Aria / `onPress`) — separate issue; this app uses `onClick`.
- Any change to `decodeTransaction.ts`, the ABI-fetching path, or `FormattedTransactionText`.
- Wiring the governance test suite into extra CI jobs beyond what `turbo test` already picks up.

## Acceptance criteria

- [ ] Zero non-null assertions on decoded args in the three files: `grep -n '!\.' apps/governance.mento.org/app/components/proposal/execution-code/patterns/token-patterns.ts apps/governance.mento.org/app/components/proposal/execution-code/patterns/proxy-patterns.ts apps/governance.mento.org/app/components/proposal/execution-code/patterns/oracle-patterns.ts` produces no output (exit code 1).
- [ ] No voter-visible error strings from the wrapper: `grep -n 'Error:' apps/governance.mento.org/app/components/proposal/execution-code/patterns/base-pattern.ts` shows no `return \`Error:` lines.
- [ ] `PatternFunction` returns `string | null` and `transaction-translator.ts` falls through to the generic description when a pattern returns `null` (verified by reading the diff; confidence for the fallthrough result stays `"medium"`, not `"high"`).
- [ ] `pnpm install` succeeds and `turbo test --filter governance.mento.org` runs the **three new test files** (verify the run summary lists 3 test files — `passWithNoTests: true` means a wrong `include` glob still exits green with 0 tests; a 0-test run is a failure of this criterion).
- [ ] Malformed-args tests (empty array, too-few args, null value) pass for every pattern function in the three files; one happy-path test per file pins a real description.
- [ ] `turbo check-types` passes (no `any` introduced; repo-wide `any` count unchanged).
- [ ] `trunk check --fix` passes.
- [ ] `grep -n 'ErrorBoundary' apps/governance.mento.org/app/components/proposal/execution-code/ExecutionCode.tsx` shows the import and usage.
- [ ] Manual runtime check on `turbo dev --filter governance.mento.org` (port 3002): open any proposal with execution code and confirm the Simple View still renders real summaries; then temporarily make `decodeTransaction` return `args: []` for one transaction and confirm the Simple View shows the generic `Call <function> on <contract>` fallback (not an `Error: ...` string); then temporarily `throw` inside `SimpleView` and confirm the boundary renders the notice + raw JSON instead of a blank card. Revert both temporary edits.

## Gotchas

- **`passWithNoTests: true`** in the shared vitest config (`packages/vitest-config/vitest.shared.js`) plus its default `include: ["src/**...]` glob is a trap: governance code lives under `app/`, so forgetting the `include` override yields a green 0-test run. Always check the reported test-file count.
- **Mock the resolver module, do not import it for real.** `address-resolver-service.ts` is `"use client"` and imports `@sentry/nextjs` and `contract-registry.json`; loading it in a node-env test drags in Sentry. `vi.mock` is hoisted above imports, so declaring it at the top of the test file is sufficient. The mock factory must export all three names the pattern files import (`getAddressNameFromCache`, `getContractInfo`, `addressResolverService`).
- **`noUncheckedIndexedAccess` in tests too:** `tokenPatterns["transfer(address,uint256)"]` is typed possibly-undefined. Use `if (!pattern) throw new Error("pattern missing")` (or `expect(pattern).toBeDefined()` plus a guard) — do not reintroduce `!` in the tests you are writing to remove `!`.
- **No `any`:** for the null-value malformed case, cast through `unknown` (`value: null as unknown as string`), never `as any`.
- **Do not add new packages.** `vitest` (`catalog:`) and `@repo/vitest-config` (`workspace:*`) already exist in the workspace; the error boundary comes from `@sentry/nextjs`, already a governance dependency. Do not add `jsdom` or `react-error-boundary`.
- **The real decoder coerces null arg values to `""`** (`utils/decodeTransaction.ts:167-182`), and stringifies uints. Model test fixtures on that (string values), and don't expect `null` back for an empty-string address — the guard targets missing/undefined args, not every odd value.
- **ErrorBoundary catches render-phase throws only.** Async translation errors in `useExecutionCodeData`/`processTransaction` are already try/caught upstream; the boundary is defence-in-depth for renderer throws, not a replacement for steps 1-3.
- Commit as `fix(governance): ...` — commitlint enforces Conventional Commits.

**Effort:** M · **Risk:** low

### LINKED ISSUE #408: test(governance): vitest bootstrap + tests for locking, quorum, and timelock logic
Part of the 2026-07 repository audit — M2 — High-leverage structure.

`apps/governance.mento.org` is the largest app in the monorepo (132 TypeScript source files) and holds real money logic: veMENTO lock maths, quorum reads, timelock operation hashing, and vote-casting transactions. It has **zero tests and no `test` script**, so `turbo test` (and the CI `pnpm test` step) silently skips it. A regression in lock-duration or timelock-hash maths would ship unnoticed. This issue bootstraps Vitest for the app (mirroring `apps/app.mento.org`) and adds behaviour tests for the four highest-value targets.

## Current behavior (evidence)

**No test script.** `apps/governance.mento.org/package.json:6-14` — the scripts block has no `test` or `test:watch`:

```json
"scripts": {
  "build": "next build --turbopack",
  "check-types": "tsc --noEmit",
  "clean": "rm -rf .next && rm -rf .turbo",
  "codegen": "graphql-codegen",
  "dev": "next dev --turbopack --port 3002",
  "lint": "npx @trunkio/launcher check",
  "start": "next start"
}
```

**Turbo and CI only run tests where a script exists.** `turbo.json:35-38`:

```json
"test": {
  "dependsOn": ["^build"],
  "cache": false
},
```

`.github/workflows/ci.yml:83-84`:

```yaml
- name: Test
  run: pnpm test
```

So adding a `test` script is sufficient for both `turbo test` and CI to pick the app up — no workflow changes needed.

**Reference setup to mirror.** `apps/app.mento.org/package.json:12-14` has the scripts:

```json
"test": "vitest run",
"test:visual": "playwright test",
"test:watch": "vitest"
```

and `apps/app.mento.org/vitest.config.ts:1-20` (both apps use the identical `@ -> ./app` tsconfig alias, so this config works for governance verbatim):

```ts
import { resolve } from "path";
import { mergeConfig } from "vitest/config";
import sharedConfig from "@repo/vitest-config/shared";

export default mergeConfig(sharedConfig, {
  esbuild: {
    jsx: "automatic",
    jsxImportSource: "react",
  },
  resolve: {
    alias: {
      // Mirror the tsconfig path alias used by the app
      "@": resolve(__dirname, "./app"),
    },
  },
  test: {
    environment: "jsdom",
    include: ["app/**/*.test.ts", "app/**/*.test.tsx"],
  },
});
```

Note the `include` override: the shared config (`packages/vitest-config/vitest.shared.js:24`) defaults to `src/**/*.test.ts`, which matches nothing in an `app/`-rooted Next.js app, and `vitest.shared.js:33` sets `passWithNoTests: true` — without the override the run passes green with zero tests.

**Test target 1 — lock maths.** `apps/governance.mento.org/app/contracts/locking/locking-helper.ts:59-71` (pure statics; `MAX_LOCKING_DURATION_WEEKS = 104` from `app/contracts/locking/config.ts:1`):

```ts
public static calculateMaxExtensionWeeks(
  currentLockingWeek: number,
  lockTime: number | undefined,
  lockSlope: number | undefined,
): number {
  if (lockTime === undefined || lockSlope === undefined) {
    return 0;
  }

  const weeksPassed = currentLockingWeek - lockTime;
  const remainingWeeks = MAX_LOCKING_DURATION_WEEKS - lockSlope;
  return Math.max(remainingWeeks + weeksPassed, 0);
}
```

plus `calculateExpirationDate` at `locking-helper.ts:73-89` (Wednesday-based week arithmetic using `startOfWeek(new Date(), { weekStartsOn: 3 })`) and date helpers such as `addYearsAndAdjustToNextWednesday` (lines 14-20) and `getDateInFutureAsWeeks` (lines 41-47, returns 0 for past dates).

**Test target 2 — quorum.** `apps/governance.mento.org/app/contracts/governor/use-quorum.ts:6-21`:

```ts
export const useQuorum = (blockNumber: bigint) => {
  const ensuredChainId = useEnsureChainId();
  const { MentoGovernor } = useContracts();

  const { data: quorumNeeded } = useReadContract({
    address: MentoGovernor.address,
    abi: GovernorABI,
    functionName: "quorum",
    args: [blockNumber],
    chainId: ensuredChainId,
  });

  return {
    quorumNeeded,
  };
};
```

**Test target 3 — timelock hashing.** `apps/governance.mento.org/app/contracts/governor/utils/get-timelock-operation-id.ts:20-38` (zero salt defined at line 8):

```ts
export function getTimelockOperationId(
  targets: readonly string[],
  values: readonly bigint[],
  calldatas: readonly `0x${string}`[],
  descriptionHash: `0x${string}`,
): `0x${string}` {
  return keccak256(
    encodePacked(
      ["address[]", "uint256[]", "bytes[]", "bytes32", "bytes32"],
      [
        targets as readonly `0x${string}`[],
        values,
        calldatas,
        descriptionHash,
        TIMELOCK_SALT,
      ],
    ),
  );
}
```

**Test target 4 — vote casting.** `apps/governance.mento.org/app/contracts/governor/use-cast-vote.tsx:39-55` wires `writeContract` with `onSuccess` (invalidates `ProposalQueryKey`) and `onError` callbacks; lines 61-67 show the user-rejection toast branch:

```ts
useEffect(() => {
  if (error) {
    if (error.message?.includes("User rejected request")) {
      toast.error("Vote transaction rejected by user");
    } else {
      toast.error("Failed to cast vote");
    }
  }
```

**Mocking pattern to copy.** `apps/app.mento.org/app/components/swap/token-dialog.test.tsx:39-42` mocks the wagmi re-export module:

```ts
vi.mock("@repo/web3/wagmi", () => ({
  useAccount: () => ({ address: undefined }),
  useChainId: () => 42220,
}));
```

See also `packages/web3/src/features/borrow/hooks/use-open-trove.test.ts` for the SDK-mocking style.

**Dependency versions in use by siblings.** `pnpm-workspace.yaml:40-41` catalogues `vitest: ^3.2.6` and `"@vitest/coverage-v8": ^3.2.4`. `apps/app.mento.org/package.json` pins `@testing-library/react: ^16.3.0` (line 53) and `jsdom: ^26.1.0` (line 60) directly (they are not in the catalog), and uses `@repo/vitest-config: workspace:*` (line 51).

## Required change

Repo rule: no new npm dependencies without approval. Every package below is already a devDependency of sibling workspaces at the same versions — adding them to governance's `package.json` **counts as approved by this issue**. Do not add anything else.

1. **`apps/governance.mento.org/package.json`** — add scripts:
   ```json
   "test": "vitest run",
   "test:watch": "vitest"
   ```
   and devDependencies (exact specifiers, matching app.mento.org):
   ```json
   "@repo/vitest-config": "workspace:*",
   "@testing-library/react": "^16.3.0",
   "@vitest/coverage-v8": "catalog:",
   "jsdom": "^26.1.0",
   "vitest": "catalog:"
   ```
   Then run `pnpm install` to update the lockfile.

2. **Create `apps/governance.mento.org/vitest.config.ts`** — copy `apps/app.mento.org/vitest.config.ts` verbatim (shown above; the `@ -> ./app` alias and `app/**` include both apply unchanged).

3. **Create `apps/governance.mento.org/app/contracts/locking/locking-helper.test.ts`** — pure-function tests, no mocks needed. Pin time with `vi.useFakeTimers()` + `vi.setSystemTime(<a known Wednesday, e.g. 2026-07-01T12:00:00Z>)` in `beforeEach`, `vi.useRealTimers()` in `afterEach` (the helpers call `new Date()` internally). Cover at least:
   - `calculateMaxExtensionWeeks`: `undefined` lockTime or lockSlope returns 0; typical values (`currentLockingWeek=10, lockTime=5, lockSlope=100` gives `104-100+5=9`); result clamped at 0 when slope exceeds remaining budget; boundary at `lockSlope === 104`.
   - `calculateExpirationDate`: known fixture — e.g. `currentWeek=10, weekLocked=8, cliff=0, slope=4` expires 2 weeks after the pinned "now"'s week start; a max-cliff case (`cliff=103, slope=1`).
   - `getDateInFutureAsWeeks`: past date returns 0; date exactly N weeks ahead returns N (floor rounding).
   - `addYearsAndAdjustToNextWednesday`: a Wednesday input is returned unchanged; a non-Wednesday input moves forward to the next Wednesday.

4. **Create `apps/governance.mento.org/app/contracts/governor/utils/get-timelock-operation-id.test.ts`** — pure hashing tests:
   - Pin a literal fixture: after `pnpm install`, compute the expected id once (e.g. `node -e` with viem's `keccak256`/`encodePacked` and the zero salt for one target/value/calldata/descriptionHash tuple), hard-code the resulting `0x…` literal in the test, and assert `getTimelockOperationId(...)` returns it. Do **not** recompute via the same code path inside the assertion — that is a tautology.
   - Assert the result is a 66-char `0x`-prefixed hex string; different `descriptionHash` yields a different id; same inputs are deterministic.

5. **Create `apps/governance.mento.org/app/contracts/governor/use-quorum.test.ts`** — hook test with `renderHook` from `@testing-library/react`. Mock (sketch — adapt to local style):
   ```ts
   vi.mock("@repo/web3", () => ({
     GovernorABI: [],
     useContracts: () => ({ MentoGovernor: { address: "0x1111111111111111111111111111111111111111" } }),
     useEnsureChainId: () => 42220,
   }));
   vi.mock("@repo/web3/wagmi", () => ({ useReadContract: vi.fn() }));
   ```
   Assert: `useReadContract` is called with `functionName: "quorum"`, `args: [blockNumber]`, the mocked governor address and chainId; `quorumNeeded` reflects the mocked `data` (e.g. `720_000n`); the zero-supply edge (`data: 0n`) is returned as `0n`, not coerced to `undefined`/falsy-dropped.

6. **Create `apps/governance.mento.org/app/contracts/governor/use-cast-vote.test.tsx`** — one wagmi-mocked hook test, happy path + error path. Mock `@repo/web3` (`GovernorABI`, `useContracts`), `@repo/web3/wagmi` (`useAccount`, `useWriteContract`, `useWaitForTransactionReceipt`), `@repo/ui` (`toast` with `success`/`error` spies), `@/hooks/use-current-chain`, and `@/contracts/governor/use-proposal` (`ProposalQueryKey`), plus `@tanstack/react-query` (`useQueryClient` returning an object with an `invalidateQueries` spy). Happy path: make the mocked `writeContract(config, options)` invoke `options.onSuccess()`; call `result.current.castVote(1n, 1, onSuccess)` inside `act(...)`; assert `writeContract` received `functionName: "castVote"` with `args: [1n, 1]`, `invalidateQueries` was called with the proposal query key, and the caller's `onSuccess` ran. Error path: make `writeContract` invoke `options.onError(error)`; assert the caller's `onError` receives it.

7. Run the verification suite: `pnpm --filter governance.mento.org test`, `turbo test --filter governance.mento.org`, `turbo check-types`, `trunk check --fix`.

## Out of scope

- Tests for any other governance modules (proposals list, execution-code decoding, GraphQL layer) — this issue establishes the harness and the four priority targets only.
- Pixel/VRT or Playwright coverage for governance.
- Refactoring the code under test (e.g. extracting the toast `useEffect` in `use-cast-vote.tsx`) — test the current behaviour as-is.
- CI workflow changes — `.github/workflows/ci.yml` already runs `pnpm test`; no edits there.
- Fixing the CLAUDE.md docs drift about React Aria/`onPress` — separate issue.

## Acceptance criteria

- [ ] `apps/governance.mento.org/package.json` has `"test": "vitest run"` and `"test:watch": "vitest"` scripts, and only the five devDependencies listed above were added — verify with `git diff main -- apps/governance.mento.org/package.json`.
- [ ] `apps/governance.mento.org/vitest.config.ts` exists and extends `@repo/vitest-config/shared` with `environment: "jsdom"` and `include: ["app/**/*.test.ts", "app/**/*.test.tsx"]` — verify with `grep -n "app/\*\*/\*.test" apps/governance.mento.org/vitest.config.ts`.
- [ ] At least 4 test files exist, colocated next to their sources: `find apps/governance.mento.org/app -name "*.test.ts*"` lists `locking/locking-helper.test.ts`, `governor/use-quorum.test.ts`, `governor/utils/get-timelock-operation-id.test.ts`, `governor/use-cast-vote.test.tsx`.
- [ ] Every test asserts observable behaviour (values, mock call arguments, callback invocations); no render-without-crash-only tests — reviewer check: `grep -rn "toBe\|toEqual\|toHaveBeenCalled" apps/governance.mento.org/app --include="*.test.*" | wc -l` is well above 4.
- [ ] `pnpm --filter governance.mento.org test` exits 0 and the Vitest summary reports **4+ test files / all passing** (not "no tests found" — `passWithNoTests` is on in the shared config).
- [ ] `turbo test --filter governance.mento.org` exits 0 (builds `@repo/web3`/`@repo/ui` first via `dependsOn: ["^build"]`).
- [ ] Root `pnpm test` output now includes a `governance.mento.org#test` task line.
- [ ] `turbo check-types` exits 0.
- [ ] `trunk check --fix` reports no issues on the changed files.
- [ ] `pnpm-lock.yaml` updated in the same commit (`pnpm install` run after the package.json edit); CI's `pnpm install` succeeds.

## Gotchas

- **Silent zero-test pass.** `packages/vitest-config/vitest.shared.js:24` includes only `src/**` and line 33 sets `passWithNoTests: true`. If you forget the `include` override in the app-local config, everything looks green while running nothing. Always confirm the file count in Vitest's summary.
- **`vi.mock` is hoisted.** Mock factories run before imports regardless of position; do not reference outer variables inside the factory without `vi.hoisted`.
- **Path-alias mocks work.** `vi.mock("@/contracts/governor/use-proposal", ...)` resolves via the `@` alias defined in `vitest.config.ts` — but only if step 2's alias block is present.
- **Time-dependent lock maths.** `LockingHelper` calls `new Date()` internally and uses Wednesday-based weeks (`weekStartsOn: 3`). Pin the system time to a known Wednesday with fake timers or the expiration/weeks tests will be flaky across week boundaries.
- **`use-cast-vote.tsx` renders JSX in a toast.** Mock `@repo/ui`'s `toast` and keep the test file `.test.tsx`; the `esbuild: { jsx: "automatic" }` block from app.mento.org's config is required for TSX under Vitest.
- **`useWaitForTransactionReceipt` must be mocked** to return `{ isLoading: false, isSuccess: false }` by default, or the hook's destructuring throws.
- **jsdom/@testing-library are not catalog entries.** Use the literal semver ranges from app.mento.org (`^26.1.0`, `^16.3.0`), and `catalog:` only for `vitest` and `@vitest/coverage-v8`; mixing these up breaks `pnpm install`.
- **`turbo test` needs built workspace deps.** Running bare `vitest` after a fresh clone fails to resolve `@repo/web3/wagmi` (it points at `dist/`); use `turbo test --filter governance.mento.org` or run `turbo build --filter @repo/web3 --filter @repo/ui` first.
- **CI also runs `pnpm knip`.** If knip flags a new devDependency as unused, mirror however app.mento.org is configured rather than deleting the dependency (app.mento.org carries the identical set and knip passes today).

**Effort:** M · **Risk:** low

### PR DESCRIPTION
**Pull request by chapati23**
Closes #408
Closes #400

## What

**#408 — Vitest bootstrap for governance.mento.org**
- Added `test`/`test:watch` scripts and the same devDependencies (`@repo/vitest-config`, `@testing-library/react`, `@vitest/coverage-v8`, `jsdom`, `vitest`) at the versions `app.mento.org` already uses.
- `vitest.config.ts` copied from `app.mento.org` (jsdom environment, `@ -> ./app` alias, `app/**/*.test.{ts,tsx}` include).
- Tests added:
  - `app/contracts/locking/locking-helper.test.ts` — pure lock-duration/expiration-date maths, pinned to a fake Wednesday system time.
  - `app/contracts/governor/utils/get-timelock-operation-id.test.ts` — hashing test against a hand-computed (not recomputed-in-place) fixture.
  - `app/contracts/governor/use-quorum.test.ts` — wagmi-mocked hook test, including the zero-supply edge case.
  - `app/contracts/governor/use-cast-vote.test.tsx` — wagmi-mocked happy-path + error-path hook test.

**#400 — Harden execution-code calldata patterns**
- `patterns/types.ts` / `patterns/base-pattern.ts`: `PatternFunction` (and `createPattern`'s `fn` param) now return `string | null`. All three `return \`Error: ...\`` sites in `createPattern` now return `null` instead of a voter-visible error string.
- `token-patterns.ts` / `proxy-patterns.ts` / `oracle-patterns.ts`: removed all 26 non-null assertions (`arg!.value`), replaced with explicit `if (x === undefined) return null;` guards right after destructuring — `noUncheckedIndexedAccess` now enforces this at compile time.
- `transaction-translator.ts`: when a pattern returns `null`, the dispatcher now falls through to the existing generic ABI-based description (`confidence: "medium"`) instead of returning a misleading `confidence: "high"` error string.
- `ExecutionCode.tsx`: wrapped the `<Tabs>` block in a Sentry `ErrorBoundary` (already used elsewhere in the app) as defence-in-depth, with a fallback that shows a short notice plus the existing `RawView` (raw calldata JSON).
- New tests: `token-patterns.test.ts`, `proxy-patterns.test.ts`, `oracle-patterns.test.ts` — malformed-args cases (empty args, too-few args, null arg value) for every guarded pattern function, plus one happy-path test per file pinning a real description.

## Why

`governance.mento.org` is the largest app in the monorepo and holds real-money logic (veMENTO lock maths, quorum reads, timelock hashing) with zero test coverage, so regressions ship unnoticed. Separately, the proposal execution-code viewer is the one screen voters use to verify what a proposal will do on-chain — its inputs are adversarial by definition, and the previous non-null assertions + swallowed-into-a-string error path meant a decoding-shape mismatch would silently render a misleading "high confidence" summary instead of falling back to the safe generic/raw view.

## Verification

```
pnpm --filter governance.mento.org exec tsc --noEmit
# exits 0, no output

pnpm --filter governance.mento.org exec vitest run
# Test Files  7 passed (7)
#      Tests  61 passed (61)

grep -n '!\.' apps/governance.mento.org/app/components/proposal/execution-code/patterns/{token,proxy,oracle}-patterns.ts
# (no output — zero non-null assertions remain)

grep -n 'Error:' apps/governance.mento.org/app/components/proposal/execution-code/patterns/base-pattern.ts
# (no output — no voter-visible error strings)

grep -n 'ErrorBoundary' apps/governance.mento.org/app/components/proposal/execution-code/ExecutionCode.tsx
# import + usage present

trunk fmt <changed files> && trunk check --fix <changed files>
# Checked 17 files — No issues

git push
# pre-push hook: turbo check-types (4/4 successful) + knip — both clean
```

## Deferred (human step)

- The manual browser check in issue #400's acceptance criteria (temporarily forcing `decodeTransaction` to return `args: []`, and temporarily throwing inside `SimpleView`, to visually confirm the fallback/boundary render) was not run interactively in this session — the behaviour is covered by the malformed-args unit tests instead. Worth a quick manual spot-check on `turbo dev --filter governance.mento.org` before merge if a visual confirmation is wanted.
- Out of scope per both issues: the 27 identical `!` assertions in `governance-patterns.ts`/`reserve-patterns.ts`/`protocol-patterns.ts`, and Playwright/VRT coverage for governance.

---

> [!NOTE]
> **Medium Risk**
> Changes affect how voters read on-chain proposal effects (summary text and error handling), though behavior is safer via fallbacks; new tests reduce regression risk for governor/locking helpers.
> 
> **Overview**
> Adds **Vitest** to `governance.mento.org` (`vitest.config.ts`, `test` scripts, jsdom) with unit tests for locking maths, timelock operation IDs, `useQuorum`, and `useCastVote`.
> 
> **Hardens the proposal execution-code viewer** so bad or partial calldata cannot surface as high-confidence summaries. Pattern helpers now return `null` instead of voter-visible `Error:` strings; token/proxy/oracle patterns drop non-null assertions in favor of explicit guards. `transaction-translator` falls through to the generic ABI description when a pattern returns `null`. `ExecutionCode` wraps the tabbed views in a Sentry `ErrorBoundary` (keyed by transaction payload) that falls back to **Raw JSON** on render failures.
> 
> New pattern tests cover malformed args for each guarded signature plus happy paths.
> 
> <sup>Reviewed by [Cursor Bugbot](https://cursor.com/bugbot) for commit 22e659d7063edbc5a4759b65a8f468eff09fb2ab. Bugbot is set up for automated code reviews on this repo. Configure [here](https://www.cursor.com/dashboard/bugbot).</sup>

### CODE REVIEW

File: apps/governance.mento.org/app/components/proposal/execution-code/ExecutionCode.tsx [hunk 1]
```diff
@@ -55,35 +56,47 @@ export function ExecutionCode({ transactions, className }: ExecutionCodeProps) {
...
 
-              <TabsContent value="raw" className="mt-6">
-                <RawView transactions={transactions} />
-              </TabsContent>
-            </Tabs>
+                <TabsContent value="raw" className="mt-6">
+                  <RawView transactions={transactions} />
+                </TabsContent>
+              </Tabs>
+            </ErrorBoundary>
```
[the remark this answers is not in the document]
Comment (chapati23) [reply apps/governance.mento.org/app/components/proposal/execution-code/ExecutionCode.tsx:106]: Fixed in 22e659d — added a transaction-derived key to the execution-code ErrorBoundary so it remounts and clears fallback state when proposal transactions change.
